### PR TITLE
DOC: optimize: document recipe for `*args` and `**kwargs`

### LINKED ISSUE #21504: DOC: add note about `args`/`kwargs` to description of callable arguments
### Is your feature request related to a problem? Please describe.

The `approx_fprime` function in `scipy.optimize` currently does not support keyword arguments (`kwargs`) for the objective function `f`. This limitation can lead to errors or warnings when `f` is a keyword-only function, as described in [PEP 3102](https://peps.python.org/pep-3102/).

PEP 3102 introduces the concept of keyword-only arguments in Python, where arguments can be passed only by keyword and not by position. This is increasingly used in modern Python code to improve clarity and prevent errors. However, when such a function is passed to `approx_fprime`, it results in errors because `approx_fprime` only supports positional arguments through `*args`.

### Describe the solution you'd like.

Enhance `approx_fprime` to accept `kwargs` and pass them to the objective function `f`. This would align `approx_fprime` with modern Python practices, reduce potential errors, and improve its flexibility.

The updated signature could look like this:

```python
def approx_fprime(xk, f, epsilon=_epsilon, *args, **kwargs):
    ...
```

## **Example of the Issue**

Consider a function that requires keyword-only arguments:

```python
import numpy as np
from scipy.optimize import approx_fprime

def func_with_kwargs(x, *, a=1.0, b=2.0):
    return a * np.sum(x**2) + b

xk = np.array([1.0, 2.0, 3.0])
epsilon = 1e-8

# This will raise an error because approx_fprime does not support kwargs
grad = approx_fprime(xk, func_with_kwargs, epsilon, 1.0, 2.0)
```

In this case, `func_with_kwargs` cannot be passed to `approx_fprime` without a workaround, such as wrapping it in another function or using `functools.partial`, both of which add unnecessary complexity.

## Benefits
- Compatibility with Modern Python Code: By supporting `kwargs`, `approx_fprime` would be compatible with functions that utilize keyword-only arguments, as per PEP 3102.

- Ease of Use: Users would no longer need to resort to workarounds like wrappers or `functools.partial` to pass keyword arguments, making the function easier to use in a broader range of scenarios.

- Backward Compatibility: The proposed change is backward-compatible, as it would not affect existing code that only uses positional arguments.

## Conclusion
Supporting `kwargs` in `approx_fprime` would make the function more versatile, aligning it with modern Python practices and improving its usability. This enhancement would be particularly beneficial for users working with keyword-only functions, which are becoming more common in the Python ecosystem.

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-21504

#### What does this implement/fix?
gh-21504 (and many others) have proposed adding `kwargs` as an additional argument to `approx_fprime`. By extension, the proposal would be to ensure that `args` and `kwargs` are accepted by all SciPy functions that accept a callable. Rather than implementing, documenting, and testing new arguments for dozens of functions, the resolution was to document (in key places) the standard recipe for users to achieve the same goal. This is already documented near the top of the `optimize` tutorial, so this adds a standard note to the `f`/`fun` argument documentation of some fundamental `optimize` functions.

#### Additional information
Because most of these functions already accept `args`, I didn't think I could use the imperative; rather, I suggest "consider wrapping the callable...". This may be misunderstood as meaning that it is entirely optional, even when there are keyword arguments. There are a few options for addressing this if it is deemed essential, but I thought that leaving it as-is was the best option.

### CODE REVIEW

**Comment by mdhaber** [comment]
To resolve the lint failure, please consider gh-21545.

**Comment by mdhaber** [comment]
@deka27 how does this look to you?

**Comment by deka27** [comment]
> @deka27 how does this look to you?

Sorry for the late reply. I think it looks good to me.

**Comment by stefanv** [comment]
This advice feels quite confusing to me. I *think* what you're suggesting is that functions are wrapped as follows:

```
f = lambda x: f0(x, arg0, arg1, foo='1', bar='2')
```

But that's not obvious when you make the signature `f0(x, *args, **kwargs)`.

`functools.partial` is another commonly used way to address this.

File: scipy/optimize/_minimize.py [hunk 1]
```diff
@@ -65,6 +65,11 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
         where ``x`` is a 1-D array with shape (n,) and ``args``
         is a tuple of the fixed parameters needed to completely
         specify the function.
+
+        Suppose the callable has signature ``f0(x, *args, **kwargs)``, where ``args``
```
Comment (mdhaber) [inline scipy/optimize/_minimize.py:69]: If you'd prefer, this could be phrased "Suppose the callable requires positional arguments besides `x` and additional keyword arguments, and these are collected into a tuple `args` and dictionary `kwargs`, respectively." But I prefer what is here to that.

File: scipy/optimize/_minimize.py [hunk 2]
```diff
@@ -65,6 +65,11 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
...
+        and ``kwargs`` are required positional and keyword arguments. Rather than
+        passing ``f0`` as the callable, consider wrapping it to accept only ``x``;
```
Comment (mdhaber) [inline scipy/optimize/_minimize.py:71]:
```suggestion
        passing ``f0`` as the callable, wrap it to accept only ``x``;
```
(And in the documentation of `args`, discourage use in new code.)

File: scipy/optimize/_minimize.py (hunk 1 above)
Comment (lucascolley) [reply scipy/optimize/_minimize.py:69]: If you are split and need a casting vote, I would vote for the rephrasing.

Comment (stefanv) [reply scipy/optimize/_minimize.py:69]: I'll try to clarify my original concern, in case it is helpful: the user is reading the function signature + docstring while getting this message, so it is quite natural to think of `*args` and `**kwargs` to pertain to the current function. Related, a common pattern (perhaps not in SciPy, but in the ecosystem) is to allow such args and kwargs as additional arguments as, e.g., `callable_args` and `callable_kwargs`. I think the approach suggested here is a perfectly fine—a better alternative, even—as long as it is made clear to the user that *they* need to gather the args, wrap a function around it, and send that function into SciPy.

Obviously, I understand what to do myself, but when reviewing / writing docstrings, I try to put on a hat of someone who is competent with Python but not familiar with unusual (or project-specific) idioms.

**Comment by stefanv** [comment]
The linter failure is due to:

```
279 |         if not isinstance(bracket, (list, tuple, np.ndarray)):
```

I think it wants `isinstance(bracket, list | tuple | np.ndarray)`. Doesn't look like an improvement, and IIRC it's slower too, so we may want to disable that update rule if possible.

**Comment by lucascolley** [comment]
The lint rule was already disabled and re-enabling it is tracked in gh-21547. It will go away if we merge main here (but I think we can just ignore it).

**Comment by stefanv** [comment]
> The lint rule was already disabled and re-enabling it is tracked in [gh-21547](https://github.com/scipy/scipy/issues/21547). It will go away if we merge main here (but I think we can just ignore it).

Thanks, Lucas, that's good to know!

File: scipy/optimize/_minimize.py (hunk 1 above)
Comment (mdhaber) [reply scipy/optimize/_minimize.py:69]: Please make an inline suggestion, such as the alternative I proposed if you prefer that. Once that is concrete, I can consider committing it. 

I understand putting on the hat of the reader... I just have a different opinion in this case.

Comment (stefanv) [reply scipy/optimize/_minimize.py:69]: I see now that we complicate things further by allowing `args`. I see three viable paths:

(1) Continue allowing args to be specified, and then say nothing more about keyword arguments, or callable wrapping.
(2) Extend `args` to have a `kwargs` dual, and change the signature to: `            fun(x, *args, **kwargs) -> float.
(3) Deprecate `args`, and describe the callable wrapping workaround in the docs.

I don't see how mixing callable wrapping and args specification can ever be anything but confusing. Easiest for now is to do (1), i.e. to simply remove the wrapping text.

Comment (mdhaber) [reply scipy/optimize/_minimize.py:69]: The wrapping text is the purpose of the PR. Please see the linked issue.

**Review by j-bowhay — APPROVED** [review]
I am happy with the changes here. I will wait for the change to be committed and to see if there is any further discussion on the open comment

File: scipy/optimize/_minimize.py [hunk 3]
```diff
@@ -65,6 +65,11 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
...
+        e.g., pass ``fun=lambda x: f0(x, *args, **kwargs)`` as the callable.
```
Comment (stefanv) [inline scipy/optimize/_minimize.py:72]: Let me try again, and see if I can clarify the confusion between `args` (in optimization function signature) and `args` in `fun` definition. Maybe I'll try more than once, and you can see which appeal to you, or not.

```suggestion
        Consider, e.g., a callable with signature ``f0(x, k, alpha=0.1)``.
        Since it has an argument and keyword argument, for the purpose of using it
        here, you'd need to wrap it to only accept a single argument, `x`.
        Pass it to this function as, e.g., ``fun=lambda x: f0(x, 3, alpha=0.1)``,
        to fix the argument and keyword argument.
```

Comment (stefanv) [inline scipy/optimize/_minimize.py:72]:
```suggestion
        Suppose the callable has signature ``f0(x, *args, **kwargs)``, where ``args``
        and ``kwargs`` are required positional and keyword arguments
        (not to be confused with the `args` argument to this function, which we
        suggest you disregard). Rather than
        passing ``f0`` as the callable, consider wrapping it to accept only ``x``;
        e.g., pass ``fun=lambda x: f0(x, *my_args, **my_kwargs)`` as the callable,
        where `args` (tuple) and `kwargs` (dict) have been gathered before 
        invoking this function.
```

Comment (stefanv) [inline scipy/optimize/_minimize.py:72]:
```suggestion
        Suppose the callable has signature ``f0(x, *args, **kwargs)``, where ``args``
        and ``kwargs`` are required positional and keyword arguments. `fun` requires
        a callable that takes a single argument, `x`, so you'll need to rewrite it,
        perhaps as follows: `fun=lambda x: f0(x, arg1, arg2, my_kwarg1=..., my_kwarg2=...)`.
        
```

Comment (mdhaber) [reply scipy/optimize/_minimize.py:72]: So I get renaming `my_args` to not conflict with `args` - how about changing it throughout? It should be the same in lines 74 and 75, at least, but why not in 69?

Comment (stefanv) [reply scipy/optimize/_minimize.py:72]: You're right; early morning!

Comment (mdhaber) [reply scipy/optimize/_minimize.py:72]: OK, so when we change `args` and `kwargs` to `my_args` and `my_kwargs`, this could become:

```suggestion
        Suppose the callable has signature ``f0(x, *my_args, **my_kwargs)``, where 
        ``my_args`` and ``my_kwargs`` are required positional and keyword arguments.
        Rather than passing ``f0`` as the callable, consider wrapping it to accept
        only ``x``; e.g., pass ``fun=lambda x: f0(x, *my_args, **my_kwargs)`` as the
        callable, where `my_args` (tuple) and `my_kwargs` (dict) have been gathered 
        before invoking this function.
```
Here, I've removed "not to be confused with the `args` argument to this function, which we suggest you disregard)" because `my_args` and `args` have different names, so it is difficult to confuse the two.

**Comment by mdhaber** [comment]
@j-bowhay please indicate which text you'd prefer to go with; i'll push the changes and let you know.

File: scipy/optimize/_minimize.py (hunk 3 above)
Comment (j-bowhay) [reply scipy/optimize/_minimize.py:72]: I am also fine with this last revision

File: scipy/optimize/_minimize.py [hunk 4]
```diff
@@ -784,6 +789,12 @@ def minimize_scalar(fun, bracket=None, bounds=None, args=(),
     fun : callable
         Objective function.
         Scalar function, must return a scalar.
+
+        Suppose the callable has signature ``f0(x, *args, **kwargs)``, where ``args``
+        and ``kwargs`` are required positional and keyword arguments. Rather than
+        passing ``f0`` as the callable, consider wrapping it to accept only ``x``;
+        e.g., pass ``fun=lambda x: f0(x, *args, **kwargs)`` as the callable.
```
Comment (mdhaber) [inline scipy/optimize/_minimize.py:796]:
```suggestion
        Suppose the callable has signature ``f0(x, *my_args, **my_kwargs)``, where
        ``my_args`` and ``my_kwargs`` are required positional and keyword arguments.
        Rather than passing ``f0`` as the callable, consider wrapping it to accept
        only ``x``; e.g., pass ``fun=lambda x: f0(x, *my_args, **my_kwargs)`` as the
        callable, where ``my_args`` (tuple) and ``my_kwargs`` (dict) have been
        gathered before invoking this function.
```

File: scipy/optimize/_optimize.py [hunk 5]
```diff
@@ -956,6 +956,11 @@ def approx_fprime(xk, f, epsilon=_epsilon, *args):
         function is an ndarray of shape (n,) (never a scalar even if n=1).
         It must return a 1-D array_like of shape (m,) or a scalar.
 
+        Suppose the callable has signature ``f0(x, *args, **kwargs)``, where ``args``
+        and ``kwargs`` are required positional and keyword arguments. Rather than
+        passing ``f0`` as the callable, consider wrapping it to accept only ``x``;
+        e.g., pass ``f=lambda x: f0(x, *args, **kwargs)`` as the callable.
```
Comment (mdhaber) [inline scipy/optimize/_optimize.py:962]:
```suggestion
        Suppose the callable has signature ``f0(x, *my_args, **my_kwargs)``, where
        ``my_args`` and ``my_kwargs`` are required positional and keyword arguments.
        Rather than passing ``f0`` as the callable, consider wrapping it to accept
        only ``x``; e.g., pass ``f=lambda x: f0(x, *my_args, **my_kwargs)`` as the
        callable, where ``my_args`` (tuple) and ``my_kwargs`` (dict) have been
        gathered before invoking this function.
```

File: scipy/optimize/_root.py [hunk 6]
```diff
@@ -31,6 +31,11 @@ def root(fun, x0, args=(), method='hybr', jac=None, tol=None, callback=None,
     ----------
     fun : callable
         A vector function to find a root of.
+
+        Suppose the callable has signature ``f0(x, *args, **kwargs)``, where ``args``
+        and ``kwargs`` are required positional and keyword arguments. Rather than
+        passing ``f0`` as the callable, consider wrapping it to accept only ``x``;
+        e.g., pass ``fun=lambda x: f0(x, *args, **kwargs)`` as the callable.
```
Comment (mdhaber) [inline scipy/optimize/_root.py:38]:
```suggestion
        Suppose the callable has signature ``f0(x, *my_args, **my_kwargs)``, where
        ``my_args`` and ``my_kwargs`` are required positional and keyword arguments.
        Rather than passing ``f0`` as the callable, consider wrapping it to accept
        only ``x``; e.g., pass ``fun=lambda x: f0(x, *my_args, **my_kwargs)`` as the
        callable, where ``my_args`` (tuple) and ``my_kwargs`` (dict) have been
        gathered before invoking this function.
```

File: scipy/optimize/_root_scalar.py [hunk 7]
```diff
@@ -71,6 +71,11 @@ def root_scalar(f, args=(), method=None, bracket=None,
     ----------
     f : callable
         A function to find a root of.
+
+        Suppose the callable has signature ``f0(x, *args, **kwargs)``, where ``args``
+        and ``kwargs`` are required positional and keyword arguments. Rather than
+        passing ``f0`` as the callable, consider wrapping it to accept only ``x``;
+        e.g., pass ``f=lambda x: f0(x, *args, **kwargs)`` as the callable.
```
Comment (mdhaber) [inline scipy/optimize/_root_scalar.py:78]:
```suggestion
        Suppose the callable has signature ``f0(x, *my_args, **my_kwargs)``, where
        ``my_args`` and ``my_kwargs`` are required positional and keyword arguments.
        Rather than passing ``f0`` as the callable, consider wrapping it to accept
        only ``x``; e.g., pass ``f=lambda x: f0(x, *my_args, **my_kwargs)`` as the
        callable, where ``my_args`` (tuple) and ``my_kwargs`` (dict) have been
        gathered before invoking this function.
```

File: scipy/optimize/_minimize.py (hunk 4 above)
Comment (stefanv) [reply scipy/optimize/_minimize.py:796]: This resolves my concern; could make it "this optimization function" if you want to be explicit that we are not talking about `f0`, but probably clear enough as is.

**Comment by j-bowhay** [comment]
thanks all

**Comment by stefanv** [comment]
Thank you for your patience with this reviewer, @mdhaber!
